### ISSUE TITLE
False negative with overloaded funtion

### ISSUE THREAD

**Issue by dmitriy-serdyuk** [python/mypy#7581]
* Bug
* A wrong implementation doesn't cause an error even though it doesn't match any of the overloaded declarations. Slightly modifying the example from the docs:
```python
from typing import Sequence, TypeVar, Union, overload

T = TypeVar('T')

class MyList(Sequence[T]):
    def __init__(self, it: T):
        self.it: T = it

    @overload
    def __getitem__(self, index: int) -> T: ...

    @overload
    def __getitem__(self, index: slice) -> Sequence[T]: ...

    def __getitem__(self, index: Union[int, slice]) -> Union[T, Sequence[T]]:
        if isinstance(index, int):
            return self.it
        elif isinstance(index, slice):
            # What if I make a typo here? Type checking should raise an error
            return self.it
        else:
            raise TypeError(...)
```
* What is the actual behavior/output?
```
Success: no issues found in 1 source file
```
* What is the behavior/output you expect?

Type error
* What are the versions of mypy and Python you are using? 0.730 and 3.6.8
* What are the mypy flags you are using? (For example --strict-optional)

I tried different ones. Same result.

**Comment by dmitriy-serdyuk** [comment]
Sorry, I misread the documentation

[Here](https://mypy.readthedocs.io/en/latest/more_types.html#type-checking-the-implementation) is the answer to my question
> implementation is type-checked against the type hints provided on the implementation.
